pkg/detectors/box/box.go

```diff
@@ -0,0 +1,133 @@
+var _ detectors.Detector = (*Scanner)(nil)
+
+var (
+	defaultClient = common.SaneHttpClient()
+	// Make sure that your group is surrounded in boundary characters such as below to reduce false positives.
+	keyPat = regexp.MustCompile(detectors.PrefixRegex([]string{"box"}) + `\b([0-9a-zA-Z]{32})\b`)
+)
+
+// Keywords are used for efficiently pre-filtering chunks.
+// Use identifiers in the secret preferably, or the provider name.
+func (s Scanner) Keywords() []string {
+	return []string{"box"}
+}
+
+func (s Scanner) Description() string {
+	return "Box is a service offering various service for secure collaboration, content management, and workflow. Box token can be used to access and interact with this data."
+}
+
+// FromData will find and optionally verify Box secrets in a given set of bytes.
+func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (results []detectors.Result, err error) {
+	dataStr := string(data)
+
+	uniqueMatches := make(map[string]struct{})
+	for _, match := range keyPat.FindAllStringSubmatch(dataStr, -1) {
+		uniqueMatches[match[1]] = struct{}{}
+	}
+
+	for match := range uniqueMatches {
+		s1 := detectors.Result{
+			DetectorType: detectorspb.DetectorType_Box,
+			Raw:          []byte(match),
+		}
+
+		if verify {
+			client := s.client
+			if client == nil {
+				client = defaultClient
+			}
+
+			isVerified, extraData, verificationErr := verifyMatch(ctx, client, match)
+			s1.Verified = isVerified
+			s1.ExtraData = extraData
+			s1.SetVerificationError(verificationErr, match)
+		}
+
+		results = append(results, s1)
+	}
+
+	return
+}
+
+func verifyMatch(ctx context.Context, client *http.Client, token string) (bool, map[string]string, error) {
+	url := "https://api.box.com/2.0/users/me"
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return false, nil, nil
+	}
+
+	req.Header = http.Header{"Authorization": []string{"Bearer " + token}}
+	req.Header.Add("content-type", "application/json")
+
+	res, err := client.Do(req)
+	if err != nil {
+		return false, nil, err
+	}
+	defer func() {
+		_, _ = io.Copy(io.Discard, res.Body)
+		_ = res.Body.Close()
+	}()
+
+	switch res.StatusCode {
+	case http.StatusOK:
+		{
+			var u user
+			if err := json.NewDecoder(res.Body).Decode(&u); err != nil {
+				return false, nil, err
+			}
+			return true, bakeExtraDataFromUser(u), nil
+		}
+	case http.StatusUnauthorized:
+		// 401 access token not found
+		// The secret is determinately not verified (nothing to do)
+		return false, nil, nil
+	default:
+		return false, nil, fmt.Errorf("unexpected HTTP response status %d", res.StatusCode)
+	}
+}
+
+func (s Scanner) Type() detectorspb.DetectorType {
+	return detectorspb.DetectorType_Box
+}
+
+func bakeExtraDataFromUser(u user) map[string]string {
+	return map[string]string{
+		"user_id":           u.ID,
+		"user_full_name":    u.Name,
+		"username":          u.Login,
+		"user_space_amount": fmt.Sprintf("%d", u.SpaceAmount),
```

Comment:
super-nit: If you want to save 20ns, you could use the more performant `strconv.FormatInt` instead of `fmt.Sprintf` 🤣. But considering @kashifkhan0771's comment, you likely won't include these, so never mind. Just a fun fact! 😅


![Screenshot 2024-10-14 at 10 01 49 AM](https://github.com/user-attachments/assets/d19af134-e10f-47ea-8aa3-c1e1f6b0229b)